source/MaterialXCore/UnitConverter.h

```diff
@@ -0,0 +1,73 @@
+//
+// TM & (c) 2019 Lucasfilm Entertainment Company Ltd. and Lucasfilm Ltd.
+// All rights reserved.  See LICENSE.txt for license.
+//
+
+/**
+* Copyright 2019 Autodesk, Inc.
+* All rights reserved.
+*
+* This computer source code and related instructions and comments are the
+* unpublished confidential and proprietary information of Autodesk, Inc. and
+* are protected under Federal copyright and state trade secret law. They may
+* not be disclosed to, copied or used by any third party without the prior
+* written consent of Autodesk, Inc.
+*
+* @file UnitConverter.h
+* @brief Unit converter classes.
+*/
+
+#ifndef MATERIALX_UNITCONVERTER_H_
+#define MATERIALX_UNITCONVERTER_H_
+
+#include <MaterialXCore/Library.h>
+
+#include <MaterialXCore/Definition.h>
+#include <MaterialXCore/Types.h>
+
+namespace MaterialX
+{
+
+class UnitConverter;
+
+/// A shared pointer to an UnitConverter
+using UnitConverterPtr = shared_ptr<UnitConverter>;
+/// A shared pointer to a const Observer
+using ConstUnitConverterPtr = shared_ptr<const UnitConverter>;
+
+/// @class UnitConverter
+/// An unit conversoin utility class.
+///
+/// An convert may be registered with a supporting document for a given UnitTypeDef.
+class UnitConverter
+{
+  public:
+    UnitConverter() { }
+    virtual ~UnitConverter() { }
+
+    /// Convert a given value in a given unit to a desired unit
+    /// @param input Input value to convert
+    /// @param inputUnit Unit of input value
+    /// @param outputUnit Unit for output value
+    virtual float convert(float input, const string& inputUnit, const string& outputUnit) const = 0;
+};
+
+
+class LengthUnitConverter : public UnitConverter
+{
+  public:
+    virtual ~LengthUnitConverter() { }
+
+    float convert(float input, const string& inputUnit, const string& outputUnit) const override;
+
+    static UnitConverterPtr create(UnitTypeDefPtr unitTypeDef);
+
+  private:
+    LengthUnitConverter(UnitTypeDefPtr unitTypeDef);
+
+    std::unordered_map<string, float> _unitScale;
```

Comment:
I'm not sure you need a derived class specialized for length units. Since we are settling for conversion using linear transforms (scale/offset), you could just add this convert logic to the UnitConverter class. Just add a map for unit offset as well and you will be able to handle conversion of _any_ unit type.

Reply:
I think in the end we decided that linear transforms are enough for this data driven unit support. And if someone ever will need non-linear transformations it can be done by an explicit conversion node instead. So I think adding the logic to UnitConverter should be enough.

Reply:
Sounds good. That's what I have now.